### PR TITLE
feat: impl protected routing bb-11

### PR DESCRIPTION
**Pull request by KeidsID**
Add a straightforward authorization plugin that responds with 401 (Unauthorized) for invalid request auth.

### CODE REVIEW

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 1]
```diff
@@ -205,6 +206,12 @@ class BaseServerApplication implements ServerApplication {
 				await this.app.register(swaggerUi, {
 					routePrefix: `/${api.version}/documentation`,
 				});
+
+				await this.app.register(authPlugin, {
+					excludedRoutePrefixes: [APIPath.AUTH],
+					// TODO [ @KeidsID ]: inject actual token verifier
+					tokenVerifier: (token) => Promise.resolve({ token }),
```
[the remark this answers is not in the document]
Comment (KeidsID) [reply apps/backend/src/libs/modules/server-application/base-server-application.ts:213]: Do I delete the entire comment or just my github tag part?

[the remark this answers is not in the document]
Comment (fshabanov) [reply apps/backend/src/libs/modules/server-application/base-server-application.ts:213]: Entire comment

**Review by revenets — APPROVED** [review]

File: apps/backend/src/libs/plugins/auth-plugin/libs/types/auth-plugin.d.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import "fastify";
+
+declare module "fastify" {
+	interface FastifyRequest {
+		user?: unknown;
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/libs/types/auth-plugin.d.ts:5]: Use proper type

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+			}
+		}
+
+		const token = request.headers.authorization;
+
+		if (!token) {
+			return await reply
+				.code(HTTPCode.UNAUTHORIZED)
+				.send({ message: "Authorization header required for this route" });
+		}
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:28]: throw AuthError instead

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+		try {
+			request.user = await tokenVerifier(token);
+		} catch {
+			reply.code(HTTPCode.UNAUTHORIZED).send({ message: "Invalid token" });
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:33]: throw AuthError instead

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		const token = request.headers.authorization;
+
+		if (!token) {
+			return await reply
+				.code(HTTPCode.UNAUTHORIZED)
+				.send({ message: "Authorization header required for this route" });
+		}
+
+		try {
+			request.user = await tokenVerifier(token);
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:31]: Token will have only userId inside, you should get user using user service

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import { type FastifyPluginCallback } from "fastify";
+
+import { HTTPCode } from "~/libs/modules/http/libs/enums/enums.js";
+import { ServerHooks } from "~/libs/plugins/libs/enums/enums.js";
+
+type PluginOptions = {
+	excludedRoutePrefixes?: string[];
+	tokenVerifier: (token: string) => Promise<unknown>;
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:8]:
```suggestion
	whiteRoutes?: string[];
	token: Token
```
Should send full token service here
Add user service here as well

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 7]
```diff
@@ -205,6 +206,11 @@ class BaseServerApplication implements ServerApplication {
 				await this.app.register(swaggerUi, {
 					routePrefix: `/${api.version}/documentation`,
 				});
+
+				await this.app.register(authPlugin, {
+					excludedRoutePrefixes: [APIPath.AUTH],
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/base-server-application.ts:211]: Create a separate constant for white routes

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { type FastifyPluginCallback } from "fastify";
+
+import { HTTPCode } from "~/libs/modules/http/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:3]:
```suggestion
import { HTTPCode } from "~/libs/modules/http/http.js";
```

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import { ServerHooks } from "~/libs/plugins/libs/enums/enums.js";
+
+type PluginOptions = {
+	excludedRoutePrefixes?: string[];
+	tokenVerifier: (token: string) => Promise<unknown>;
+};
+
+const authPlugin: FastifyPluginCallback<PluginOptions> = (
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:11]: Use `fp` from `fastify-plugin`

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	{ excludedRoutePrefixes = [], tokenVerifier },
+) => {
+	app.addHook(ServerHooks.PRE_HANDLER, async (request, reply) => {
+		for (const routePrefix of excludedRoutePrefixes) {
+			if (request.url.startsWith(routePrefix)) {
+				return;
+			}
+		}
+
+		const token = request.headers.authorization;
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:22]:
```suggestion
		const token = request.headers[HTTPHeader.AUTHORIZATION];
```

File: apps/backend/src/libs/plugins/libs/enums/hooks.enum.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+const ServerHooks = {
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/libs/enums/hooks.enum.ts:1]: File name should be the same with variable name

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 12]
```diff
@@ -5,12 +5,13 @@ import Fastify, { type FastifyError, type FastifyInstance } from "fastify";
...
 import { fileURLToPath } from "node:url";
 
-import { ServerErrorType } from "~/libs/enums/enums.js";
+import { APIPath, ServerErrorType } from "~/libs/enums/enums.js";
 import { type ValidationError } from "~/libs/exceptions/exceptions.js";
 import { type Config } from "~/libs/modules/config/config.js";
 import { type Database } from "~/libs/modules/database/database.js";
 import { HTTPCode, HTTPError } from "~/libs/modules/http/http.js";
 import { type Logger } from "~/libs/modules/logger/logger.js";
+import { authPlugin } from "~/libs/plugins/auth-plugin/auth-plugin.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/base-server-application.ts:14]:
```suggestion
import { authPlugin } from "~/libs/plugins/plugins.js";
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+const authPlugin: FastifyPluginCallback<PluginOptions> = (
+	app,
+	{ excludedRoutePrefixes = [], tokenVerifier },
+) => {
+	app.addHook(ServerHooks.PRE_HANDLER, async (request, reply) => {
+		for (const routePrefix of excludedRoutePrefixes) {
+			if (request.url.startsWith(routePrefix)) {
+				return;
+			}
+		}
```
Comment (WilhelmWesser) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:20]: Let's reduce nesting here with the following construction
```
const urlThatStartsWithRoutePrefix = excludedRoutePrefixes.find(routePrefix => request.url.startsWith(routePrefix))
if (urlThatStartsWithRoutePrefix) {
return
}
```

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts (hunk 3 above)
Comment (WilhelmWesser) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:28]: Maybe let's move the error message to a separate enum?

Comment (WilhelmWesser) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:28]: Maybe let's move the error message to a separate enum?

Comment (KeidsID) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:28]: Can't find `AuthError`. Do I need to create it or just use the `HTTPError` from "shared" lib?

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts (hunk 6 above)
Comment (KeidsID) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:8]: Do I need to pick commits from PR #40 or create a `Token` abstraction based on that PR?

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts (hunk 3 above)
Comment (fshabanov) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:28]: Check #37, AuthError was added there. You should add it as well

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts (hunk 6 above)
Comment (fshabanov) [reply apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:8]: You can do that, yes

File: apps/backend/src/libs/modules/server-application/libs/constants/constants.ts [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import { APIPath, AuthApiPath } from "shared";
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/constants/constants.ts:1]: Don't import directly from shared folder

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 15]
```diff
@@ -205,6 +209,12 @@ class BaseServerApplication implements ServerApplication {
 				await this.app.register(swaggerUi, {
 					routePrefix: `/${api.version}/documentation`,
 				});
+
+				await this.app.register(authPlugin, {
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/base-server-application.ts:213]:
```suggestion
				await this.app.register(authorizationPlugin, {
```

File: apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+			if (whiteRoute) {
+				return;
+			}
+
+			const header = request.headers[HTTPHeader.AUTHORIZATION];
+
+			if (!header) {
+				throw new AuthError({
+					message: "Authorization header required for this route",
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/auth-plugin/auth-plugin.ts:31]: Message should come from enum

File: apps/backend/src/modules/users/user.repository.ts [hunk 17]
```diff
@@ -38,6 +38,17 @@ class UserRepository implements Repository {
 		return users.map((user) => UserEntity.initialize(user));
 	}
 
+	public async findByEmail(email: string): Promise<null | UserEntity> {
+		const user = await this.userModel.query().findOne({ email });
+		return user ? UserEntity.initialize(user) : null;
+	}
+
+	public async findById(id: string): Promise<null | UserEntity> {
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.repository.ts:46]: For finding by id we should use existing `find` method

File: apps/backend/src/modules/users/user.service.ts [hunk 18]
```diff
@@ -45,6 +43,18 @@ class UserService implements Service {
 		};
 	}
 
+	public async findByEmail(email: string): Promise<null | UserDto> {
+		const user = await this.userRepository.findByEmail(email);
+
+		return user ? user.toObject() : null;
+	}
+
+	public async findById(id: string): Promise<null | UserDto> {
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.service.ts:52]: Same here

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/libs/enums/enums.ts [hunk 19]
```diff
@@ -1,2 +1,8 @@
 export { RelationName } from "./relation-name.enum.js";
-export { APIPath, AppEnvironment, ErrorMessage, ServerErrorType } from "shared";
+export {
+	APIPath,
+	AppEnvironment,
+	AuthApiPath,
```
Comment (fshabanov) [inline apps/backend/src/libs/enums/enums.ts:5]: Should be exported from auth module

File: apps/backend/src/libs/exceptions/exceptions.ts [hunk 20]
```diff
@@ -1 +1 @@
-export { ValidationError } from "shared";
+export { AuthError, ValidationError } from "shared";
```
Comment (fshabanov) [inline apps/backend/src/libs/exceptions/exceptions.ts:1]: Same, from auth module

File: apps/backend/src/libs/plugins/authorization-plugin/libs/types/enums/authorization-plugin-error-message.enum.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
+const AuthorizationPluginErrorMessage = {
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/libs/types/enums/authorization-plugin-error-message.enum.ts:1]: All error messages should be in enum, that we have in shared package

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+				}
+
+				throw new AuthError({
+					cause: error,
+					message: ErrorMessage.INVALID_TOKEN,
+				});
+			}
+		});
+	},
+	{ name: "authorization-plugin" },
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:62]:
```suggestion
```

File: apps/backend/src/libs/types/repository.type.ts [hunk 23]
```diff
@@ -1,7 +1,7 @@
 type Repository<T = unknown> = {
 	create(payload: unknown): Promise<T>;
 	delete(): Promise<boolean>;
-	find(): Promise<T>;
+	find(query: unknown): Promise<T>;
```
Comment (fshabanov) [inline apps/backend/src/libs/types/repository.type.ts:4]: `find` can only accept id

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/libs/enums/enums.ts [hunk 24]
```diff
@@ -1,2 +1,3 @@
 export { RelationName } from "./relation-name.enum.js";
+export { AuthApiPath } from "~/modules/auth/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/enums/enums.ts:2]: Shouldn't be exported from ~/libs folder

File: apps/backend/src/libs/exceptions/exceptions.ts [hunk 25]
```diff
@@ -1 +1,2 @@
+export { AuthError } from "~/modules/auth/libs/exceptions/exceptions.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/exceptions/exceptions.ts:1]: Shouldn't be exported from ~/libs folder

**Review by fshabanov — CHANGES_REQUESTED** [review]
Pull the latest changes

File: apps/backend/src/libs/modules/server-application/libs/enums/enums.ts [hunk 26]
```diff
@@ -0,0 +1,2 @@
+export { APIPath } from "~/libs/enums/enums.js";
+export { AuthApiPath } from "~/modules/auth/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/enums/enums.ts:2]: Shouldn't be exported from here. Auth files should only be exported from auth module. And imported from auth module as well

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 27]
```diff
@@ -0,0 +1,64 @@
+import fp from "fastify-plugin";
+
+import { HTTPHeader } from "~/libs/modules/http/http.js";
+import { BaseToken } from "~/libs/modules/token/base-token.module.js";
+import { ServerHooks } from "~/libs/plugins/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:5]:
```suggestion
import { ServerHooks } from "../libs/enums/enums.js";
```

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 28]
```diff
@@ -0,0 +1,64 @@
+import fp from "fastify-plugin";
+
+import { HTTPHeader } from "~/libs/modules/http/http.js";
+import { BaseToken } from "~/libs/modules/token/base-token.module.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:4]:
```suggestion
import { BaseToken } from "~/libs/modules/token/token.js";
```

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+import { ServerHooks } from "~/libs/plugins/libs/enums/enums.js";
+import { TokenPayload } from "~/libs/types/types.js";
+import { UserService } from "~/modules/users/user.service.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:7]:
```suggestion
import { UserService } from "~/modules/users/users.js";
```

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+
+import { AuthError, ErrorMessage } from "./libs/types/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:9]: You don't need enums folder here. `import { AuthError } from ~/modules/auth/auth.js` and `import { ErrorMessage } from ~/libs/enums/enums.js`

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+			if (whiteRoute) {
+				return;
+			}
+
+			const header = request.headers[HTTPHeader.AUTHORIZATION];
+
+			if (!header) {
+				throw new AuthError({
+					cause: request,
+					message: ErrorMessage.MISSING_AUTHORIZATION_HEADER,
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:33]: Just send Unauthorized, users doesn't need more info

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 32]
```diff
@@ -0,0 +1,64 @@
...
+
+			if (whiteRoute) {
+				return;
+			}
+
+			const header = request.headers[HTTPHeader.AUTHORIZATION];
+
+			if (!header) {
+				throw new AuthError({
+					cause: request,
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:32]:
```suggestion
```

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 33]
```diff
@@ -0,0 +1,64 @@
...
+
+			try {
+				const {
+					payload: { userId },
+				} = await token.decode(header);
+
+				const user = await userService.find(userId);
+
+				if (!user) {
+					throw new AuthError({ message: ErrorMessage.USER_NOT_FOUND });
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:45]: Send Unauthorized

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+
+				request.user = user.toObject();
+			} catch (error) {
+				if (error instanceof AuthError) {
+					throw error;
+				}
+
+				throw new AuthError({
+					cause: error,
+					message: ErrorMessage.INVALID_TOKEN,
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:56]: Send Unauthorized

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+				}
+
+				request.user = user.toObject();
+			} catch (error) {
+				if (error instanceof AuthError) {
+					throw error;
+				}
+
+				throw new AuthError({
+					cause: error,
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:55]:
```suggestion
```

File: apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts (hunk 22 above)
Comment (fshabanov) [reply apps/backend/src/libs/plugins/authorization-plugin/authorization-plugin.ts:61]: Not resolved

File: apps/backend/src/libs/plugins/authorization-plugin/libs/types/enums/enums.ts [hunk 36]
```diff
@@ -0,0 +1,3 @@
+export { AuthApiPath } from "~/modules/auth/libs/enums/enums.js";
+export { AuthError } from "~/modules/auth/libs/exceptions/exceptions.js";
+export { ErrorMessage } from "shared";
```
Comment (fshabanov) [inline apps/backend/src/libs/plugins/authorization-plugin/libs/types/enums/enums.ts:3]: No need to export from here

File: apps/backend/src/libs/types/service.type.ts [hunk 37]
```diff
@@ -1,7 +1,7 @@
 type Service<T = unknown> = {
 	create(payload: unknown): Promise<T>;
 	delete(): Promise<boolean>;
-	find(): Promise<T>;
+	find(query: unknown): Promise<T>;
```
Comment (fshabanov) [inline apps/backend/src/libs/types/service.type.ts:4]: `find` can only receive id

File: apps/backend/src/modules/auth/auth.controller.ts [hunk 38]
```diff
@@ -84,6 +85,13 @@ class AuthController extends BaseController {
 	 *                    type: string
 	 *                    description: "Authentication token for the user."
 	 *                    example: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ..."
+	 *                  message:
+	 *                    type: object
+	 *                    $ref: "#/components/schemas/User"
+	 *                  token:
+	 *                    type: string
+	 *                    description: "Authentication token for the user."
+	 *                    example: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ..."
```
Comment (fshabanov) [inline apps/backend/src/modules/auth/auth.controller.ts:94]:
```suggestion
```

File: apps/backend/src/modules/users/user.repository.ts [hunk 39]
```diff
@@ -53,8 +53,24 @@ class UserRepository implements Repository {
 		return Promise.resolve(true);
 	}
 
-	public find(): ReturnType<Repository["find"]> {
-		return Promise.resolve(null);
+	public async find(id: number): Promise<null | UserEntity> {
+		const user = await this.userModel
+			.query()
+			.withGraphFetched(RelationName.USER_DETAILS)
+			.where({ id })
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.repository.ts:60]: Use findById

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by WilhelmWesser — APPROVED** [review]

File: apps/backend/src/libs/modules/server-application/libs/constants/constants.ts [hunk 40]
```diff
@@ -0,0 +1,10 @@
+import { AuthApiPath } from "~/modules/auth/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/constants/constants.ts:1]:
```suggestion
import { AuthApiPath } from "~/modules/auth/auth.js";
```

File: apps/backend/src/libs/modules/server-application/libs/constants/constants.ts [hunk 41]
```diff
@@ -0,0 +1,10 @@
...
+
+import { APIPath } from "../enums/enums.js";
+
+const WHITE_ROUTES: string[] = [
+	APIPath.AUTH,
+	`${APIPath.AUTH}${AuthApiPath.SIGN_UP}`,
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/constants/constants.ts:7]: Add sign in as well

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by fshabanov — APPROVED** [review]

**Pull request merged**